extensions/mssql/src/sqlNotebook/sqlNotebookController.ts

```diff
@@ -0,0 +1,316 @@
+			throw new Error(localize('queryInProgressError', 'Another query is currently in progress. Please wait for that query to complete before running these cells.'));
+		}
+
+		let connection = this._connectionsMap.get(notebook.uri);
+		if (!connection) {
+			connection = await this.changeConnection(notebook);
+		}
+
+		let executionOrder = this._executionOrderMap.get(notebook.uri) ?? 0;
+		for (let cell of cells) {
+			await this.doExecution(cell, connection, ++executionOrder);
+		}
+		this._executionOrderMap.set(notebook.uri, executionOrder);
+	}
+
+	private async doExecution(cell: vscode.NotebookCell, connection: azdata.connection.Connection | undefined, executionOrder: number): Promise<void> {
+		const execution = this._controller.createNotebookCellExecution(cell);
+		execution.executionOrder = executionOrder;
+		execution.start(Date.now());
+		await execution.clearOutput();
+		if (!connection) {
+			await execution.appendOutput([
+				new vscode.NotebookCellOutput([
+					vscode.NotebookCellOutputItem.text(localize('noConnectionError', 'No connection provided.'))
+				])
+			]);
+			execution.end(false, Date.now());
+			return;
+		}
+
+		let cancelHandler: vscode.Disposable;
+		try {
+			const ownerUri = await azdata.connection.getUriForConnection(connection.connectionId);
+			await this._queryProvider.runQueryString(ownerUri, cell.document.getText());
+			cancelHandler = execution.token.onCancellationRequested(async () => await this._queryProvider.cancelQuery(ownerUri));
+
+			let queryComplete = new Promise<void>(resolve => {
+				let queryCompleteHandler = async (batchSummaries: azdata.BatchSummary[]) => {
+					let tableHtmlEntries: string[] = [];
+					for (let batchSummary of batchSummaries) {
+						if (execution.token.isCancellationRequested) {
+							break;
+						}
+
+						for (let resultSummary of batchSummary.resultSetSummaries) {
+							if (execution.token.isCancellationRequested) {
+								break;
+							}
+
+							if (resultSummary.rowCount > 0) {
+								// Add column headers
+								let tableHtml =
+									`<style>
+										.output_container .sqlNotebookResults td, .output_container .sqlNotebookResults th {
+											text-align: left;
+										}
+									</style>`;
+								tableHtml += '<table class="sqlNotebookResults"><thead><tr>';
+								for (let column of resultSummary.columnInfo) {
+									tableHtml += `<th>${column.columnName}</th>`;
+								}
+								tableHtml += '</tr></thead>';
+
+								// Add rows and cells
+								let subsetResult = await this._queryProvider.getQueryRows({
+									ownerUri: ownerUri,
+									batchIndex: batchSummary.id,
+									resultSetIndex: resultSummary.id,
+									rowsStartIndex: 0,
+									rowsCount: resultSummary.rowCount
+								});
+								tableHtml += '<tbody>';
+								for (let row of subsetResult.resultSubset.rows) {
+									tableHtml += '<tr>';
+									for (let cell of row) {
+										tableHtml += `<td>${cell.displayValue}</td>`;
```

Comment:
Is this being properly escaped? (same with the column names and anything else we get from the user)